app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -0,0 +1,255 @@
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Return to previous screen");
+                finish();
+            }
+        });
+
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onActivityResult(final int requestCode, final int resultCode, final Intent intent) {
+        final IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
+        if (scanResult == null) {
+            LOGGER.error("No scan result in QR Scanner");
+        } else {
+            final String result = scanResult.getContents();
+            addNewContact(result);
+        }
+
+
+    }
+
+
+    /**
+     * Adds new contact with the scanned key and opens change
+     *
+     * @param qrMessage The information we got from the QR code.
+     */
+    private void addNewContact(final String qrMessage) {
+        try {
+            LOGGER.info("Adding new contact to database");
+            //TODO: Add recognizer for contact name to avoid saving the same key twice (add your personal name to QR code)
+            final IKey key = QRGenerator.deconstructToKey(qrMessage);
+            final EditText editName = new EditText(this);
+            editName.setInputType(InputType.TYPE_CLASS_TEXT);
+            final EditNameClickListener enClickListener = new EditNameClickListener(editName, key);
+            final AlertDialog.Builder builder = new AlertDialog.Builder(this)
+                    .setTitle(getString(R.string.contact_set_name))
+                    .setView(editName)
+                    .setPositiveButton(getString(R.string.popup_done), enClickListener);
+            lastDialog = builder.create();
+            lastDialog.show();
+        } catch (IllegalArgumentException e) {
+            LOGGER.error("Illegal argument exception in addNewContact", e);
+        }
+    }
+
+    /**
+     * returns last created alert dialog
+     * @return the last AlertDialog.
+     */
+    public AlertDialog getLastDialog() {
+        return lastDialog;
+    }
+
+
+
+    /**
+     * Shows the QR Code in an alert dialog popup screen
+     *
+     * @param qrCode The QR code to be shown.
+     */
+    private void showQRCode(final Bitmap qrCode) {
+
+        final LayoutInflater li = LayoutInflater.from(this);
+        final View myView = li.inflate(R.layout.qrcode, null);
+
+        final ImageView imageView = (ImageView) myView.findViewById(R.id.QRCodeImage);
+        imageView.setImageBitmap(qrCode);
+        final AlertDialog.Builder builder = new AlertDialog.Builder(this)
+                .setView(imageView)
+                .setPositiveButton(getString(R.string.popup_done), new QRCloser());
+
+        ((ViewGroup) imageView.getParent()).removeView(imageView);
+        lastDialog = builder.create();
+        lastDialog.show();
+    }
+
+    /**
+     * Gets the public key of the user
+     * @return the public key.
+     */
+    public IKey getPublicKey() {
```

Comment:
You shouldn't be adding methods you need for testing, you can use the `BaseTest` class to access private attributes 😉